napari/plugins/io.py

```diff
@@ -74,3 +74,98 @@ def read_data_with_plugins(
                 # error anyway, so it looks a bit weird to show them that the
                 # "builtin plugin" didn't work.
                 logger.error(err.format_with_contact_info())
+
+
+def write_data_with_plugins(
+    path: str, layer_data: LayerData, plugin_manager=None
+):
+    """Iterate writer hooks and write data with first successful writer.
+
+    This function returns as soon as the data has been written successfully,
+    while catching any plugin exceptions, storing them for later retrievial,
+    providing useful error messages, and relooping until either data is
+    writen, or no valid writers are found.
+
+    Exceptions will be caught and stored as PluginErrors
+    (in plugin_manager._exceptions)
+
+    Parameters
+    ----------
+    path : str
+        The path (file, directory, url) to save.
+    layer_data : LayerData
+        ``LayerData`` is a list tuples, where each tuple is
+         ``(data, meta, layer_type)``.
+    plugin_manager : pluggy.PluginManager, optional
+        Instance of a pluggy PluginManager.  by default the main napari
+        plugin_manager will be used.
+    """
+    layer_types = [single_layer_data[2] for single_layer_data in layer_data]
+    plugin_manager = plugin_manager or napari_plugin_manager
+    skip_impls = []
+    while True:
+        (writer, implementation) = execute_hook(
+            plugin_manager.hook.napari_get_writer,
+            path=path,
+            layer_types=layer_types,
+            return_impl=True,
+            skip_impls=skip_impls,
+        )
+        if not writer:
+            # we're all out of writer plugins
+            return None
+        try:
+            return writer(path, layer_data)  # try to write data
+        except Exception as exc:
+            # If execute_hook did return a writer, but the writer then failed
+            # while trying to write the path, we store the traceback for later
+            # retrieval, warn the user, and continue looking for writers
+            # (skipping this one)
+            msg = (
+                f"Error in plugin '{implementation.plugin_name}', "
+                "hook 'napari_get_writer'"
+            )
+            # instantiating this PluginError stores it in
+            # plugins.exceptions.PLUGIN_ERRORS, where it can be retrieved later
+            err = PluginError(
+                msg, implementation.plugin_name, implementation.plugin.__name__
+            )
+            err.__cause__ = exc  # like `raise PluginError() from exc`
+
+            skip_impls.append(implementation)  # don't try this impl again
+            if implementation.plugin_name != 'builtins':
+                # If builtins doesn't work, they will get a "no writer" found
+                # error anyway, so it looks a bit weird to show them that the
+                # "builtin plugin" didn't work.
+                logger.error(err.format_with_contact_info())
+
+
+def write_image_with_plugin(
```

Comment:
@tlambert03 - curious if you think this function should be `write_layer_with_plugin` and accept a `layer_type: str` arg and then call the appropriate hook based on that arg instead of having one of these for each of our layer types?

Reply:
Another way would be for this function to just accept the layer object and to do the getting of the data and metadata dictionary inside this function. Is that better?

Reply:
I like the `write_layer_...` function that you've started... and I do actually like the idea of just passing a layer, and making sure all layers provide the same interface for retrieving meta